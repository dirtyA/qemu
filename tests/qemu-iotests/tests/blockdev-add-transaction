#!/usr/bin/env python3
#
# Test blockdev-add transaction action
#
# Copyright (c) 2022 Virtuozzo International GmbH.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import iotests
from iotests import log

iotests.script_initialize(supported_fmts=['qcow2'])

with iotests.VM() as vm:
    vm.launch()

    # Use same node-name for nodes, neither one should appear.
    vm.qmp_log('transaction', actions=[
        {'type': 'blockdev-add',
         'data': {'node-name': 'node0', 'driver': 'null-co',
                  'size': 1024 * 1024}},
        {'type': 'blockdev-add',
         'data': {'node-name': 'node0', 'driver': 'null-co',
                  'size': 1024 * 1024}}
    ])

    n = len(vm.qmp('query-named-block-nodes')['return'])
    log(f'Created {n} nodes')

    vm.qmp_log('transaction', actions=[
        {'type': 'blockdev-add',
         'data': {'node-name': 'node0', 'driver': 'null-co',
                  'size': 1024 * 1024}},
        {'type': 'blockdev-add',
         'data': {'node-name': 'node1', 'driver': 'null-co',
                  'size': 1024 * 1024}}
    ])

    n = len(vm.qmp('query-named-block-nodes')['return'])
    log(f'Created {n} nodes')
