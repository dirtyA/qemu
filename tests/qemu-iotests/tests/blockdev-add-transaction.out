{"execute": "transaction", "arguments": {"actions": [{"data": {"driver": "null-co", "node-name": "node0", "size": 1048576}, "type": "blockdev-add"}, {"data": {"driver": "null-co", "node-name": "node0", "size": 1048576}, "type": "blockdev-add"}]}}
{"error": {"class": "GenericError", "desc": "Duplicate nodes with node-name='node0'"}}
Created 0 nodes
{"execute": "transaction", "arguments": {"actions": [{"data": {"driver": "null-co", "node-name": "node0", "size": 1048576}, "type": "blockdev-add"}, {"data": {"driver": "null-co", "node-name": "node1", "size": 1048576}, "type": "blockdev-add"}]}}
{"return": {}}
Created 2 nodes
